redwood-treehouse-guest/src/jsMain/kotlin/app/cash/redwood/treehouse/ProtocolBridgeJs.kt

```diff
@@ -138,12 +138,20 @@ internal class FastGuestProtocolAdapter(
     changes.push(js("""["property",{"id":id,"tag":tag,"value":value}]"""))
   }
 
-  override fun appendModifierChange(
-    id: Id,
-    elements: List<ModifierElement>,
-  ) {
+  override fun appendModifierChange(id: Id, value: Modifier) {
+    val elements = js("[]")
+
+    value.forEach { element ->
+      val tag = widgetSystemFactory.modifierTag(element)
+      val serializer = widgetSystemFactory.modifierSerializer(element)
+      val value = when {
+        serializer == null -> null
+        else -> json.encodeToDynamic(serializer, element)
+      }
+      elements.push(js("""[tag,value]"""))
```

Comment:
The protocol when used normally does not encode `null` values here. There's no real harm in it, though, except for the 5 extra bytes on the wire.

Reply:
We might need to move our protocol serialization tests upward to ensure these custom forms mirror it in behavior.

Reply:
Ooooh good catch, I missed that. I think it’s a nice improvement to test at the JSON-encoded layer for this, though I don’t think Zipline’s `asDynamicFunction()` makes this particularly easy. It’s trying to hide the JSON layer and the test wants to exercise it!

Reply:
Fixed to not encode null, but I haven’t yet fixed the test case to validate the JSON matches as specified. I’ll need to contemplate how to do that in a way that feels right at this layer.